### DISCUSSION TITLE
Function 'button_1' repeating without pressing button  - Using Micropython and Pi Pico

### DISCUSSION THREAD

**Discussion by floordoc** [micropython/micropython#11727] [RP2040 / Pico]
This is my first post to the forum. I have found helpful posts here before, but have not found any posts similar to the problem I'm currently having. I've tried to include all of the relevant information including, the hardware being used, as well as the full code we are trying to develop. However, if things are missing, let me know. New to Python coding so the program is certainly not the most efficient. 

Within the posted program, 'button_1' does not run until the button is pressed the first time. But then, repeats without button presses.

For the hardware setup we have an IR Reflectivity sensor connected to the Pi Pico at pins (ground@pin 35, 3V3out@pin36, pin35, GP28@pin 34). We have a SPST Normally open momentary push button connected to pins (ground@pin18, gp14@pin19). We have a 16x2 display connected at (Vbus@pin40, gp0@pin1, gp1@pin2, ground@pin3). 

We are incrementing a  counter 'drop' for each button press. The drop counter goes from 1-4. When drop =4 we are computing 1 of 2 averages. We are also incrementing a 'test' counter. The test increments when the drop=4, at which time drop is returned to 1. 

```py
from machine import I2C, Pin
from utime import sleep
import time
import sys

# Sets up button
button1 = machine.Pin(14, machine.Pin.IN, machine.Pin.PULL_UP)

# Sets sensor pin
ir=Pin(28,Pin.IN)

# Runs screen
from pico_i2c_lcd import I2cLcd
i2c = I2C(0, sda=Pin(0), scl=Pin(1), freq=400000)
I2C_ADDR = i2c.scan()[0]
lcd = I2cLcd(i2c, I2C_ADDR, 2, 16)

def testing():
    global text
    i = 0
    while i < 10:
        try:
            obstacle=not(ir.value())
            
            if obstacle:					# Checks if the sensor sees anything
                if i == 0:					# The first time the sensor sees something, prints "raise weight"
                    i = 1
                    lcd.clear()
                    lcd.putstr(text)
                    lcd.putstr("Raise Weight")
                elif i == 2:				# Records the beginning of the drop
                    i = 3
                    d1 = time.ticks_us()
                elif i == 4:				# Records the beginning of the rebound
                    i = 5
                    r1 = time.ticks_us()
                    
            else:							# If the sensor sees nothing
                if i == 1:					# once weight is lifted, prints "ready to drop"
                    lcd.clear()
                    lcd.putstr(text)
                    lcd.putstr("Ready to Drop")
                    i=2
                elif i == 3:				# Records the end of the drop, and computes the drop time
                    d2 = time.ticks_us()
                    t1 = int(time.ticks_diff(d2,d1))
                    i = 4
                elif i == 5:				# Records the end of the rebound, and computes rebound time and energy return
                    r2 = time.ticks_us()
                    t2 = int(time.ticks_diff(r2,r1))
                    energy = 1 - ((t1/t2)**2)
                    return ((t1/t2)**2)		# Returns energy return
                    i = 12
            
        except OSError as e:
            machine.reset()



test = 0
drop = 1
data = [[0]*4]*4
functionStart = 1


# Cycles through tests as the button is pressed
def button_1(pin):
    global test
    global drop
    global data
    global functionStart
    print(functionStart)
    functionStart+=1
    if drop == 4:					# Calculates and displays averages as "drop 4" (Averages are stored in column 0)
        if test == 0:				# Calibration averages are taken from all 3 drops
            data[0][0] = (data[0][1] + data[0][2] + data[0][3])/3
            lcd.clear()
            lcd.putstr("Calib Average\n")
            lcd.putstr(str(data[0][0]))
            time.sleep(2)
        else:						# Test averages are taken from drops 2 and 3
            data[test][0] = (data[test][2]+data[test][3])/2
            text = "Test "+str(test)+"Average\n"
            lcd.clear()
            lcd.putstr(text)
            lcd.putstr(str(data[test][0]))
            time.sleep(2)
        drop = 1
        test+=1
    elif test == 0:
        # Run Calibration
        calib()
        drop+=1
    elif test > 0 & test < 4:
        # Run Rim Test
        rim()
        drop+=1
    else:
        # Display Values
        a = 1
        print(a)
    print("Function Reached its End")
    return None
        

def calib():
    global drop
    global data
    global text
    text = "Calib Test "+str(drop)+"\n"
    lcd.clear()
    lcd.putstr(text)
    data[0][drop] = 1-testing()				# Runs a test, calculates cal 1
    lcd.clear()
    lcd.putstr(text)
    lcd.putstr(str(data[0][drop]))
    time.sleep(2)
    return None
    
def rim():
    global drop
    global data
    global text
    global test
    text = "Test "+str(test)+" Drop "+str(drop)+"\n"
    lcd.clear()
    lcd.putstr(text)
    data[test][drop] = 1-testing()-data[0][0]				# Runs a test, calculates force absorbed
    lcd.clear()
    lcd.putstr(text)
    lcd.putstr(str(data[test][drop]))
    time.sleep(2)
    return None

print("Rim Test Device")
lcd.clear()
lcd.putstr("ASET Services\n")
lcd.putstr("Rim Test Device\n")
time.sleep(2)
avgCal = 0

lcd.clear()
lcd.putstr("Press Button 1\nto Start")

button1.irq(trigger = machine.Pin.IRQ_FALLING, handler = button_1)
```

**Comment by 2dof** [comment]
"(...)But then, repeats without button presses."  do you mean you press button and it repeat handler function couple times ? if yes then it is caused by mechanical button contacts vibration : bouncing effect)    and a some debouncing must be implemented (software or mechanical), but best way is to use [Pushbutton class](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/DRIVERS.md#41-pushbutton-class)  it is async and already have debouncing and press and release event functions.

**Reply by floordoc** [reply]
In debugging the system we tried the pushbutton part of this program in a simpler version. One that when pressed, a count incremented by 1 and was displayed. That version registered very few double button event, and no 4, 5, or 6 button events. We will look into your suggestion. For now, we changed to run the program in a while loop with button presses to advance the stages of the loop, rather than looking for state changes. That worked without registering double button presses. Same button and connections to the pi pico.

**Comment by peterhinch** [comment]
@floordoc I suggest you read [this official doc](http://docs.micropython.org/en/latest/reference/isr_rules.html) on how to write an interrupt service routine (ISR). Hard interrupts are not the best way to interface mechanical switches and complex ISR's are bad practice.

**Reply by floordoc** [reply]
I will look at this. Thank you.
